### PR TITLE
Fix jit outplace tracing and reapply changes to *_like operators.

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Reapply reverted and fix files `gen_variable_type.py` `test_jit.py`

#27891 Cleanup testing of _like operators
#27890 Add memory format support to randn_like operator
#27889 Add memory format support to randint_like operator
#27562 Add memory format support to zeros_like operator
#27561 Add memory format support to rand_like operator
#27270 Add memory format support to ones_like operator
#27262 Add memory format support to full_like operator

### CODE REVIEW

**Comment by ezyang** [comment]
Soooo... what's the fix?

File: tools/autograd/gen_variable_type.py [hunk 1]
```diff
@@ -48,8 +48,22 @@
...
 RENAME_TRACE = {
-    'zero': 'zeros_like',
-    'fill': 'full_like',
+    'zero': 'zeros_like',  # replacing aten::zero_ with aten::zeros_like
+    'fill': 'full_like',  # replacing aten::fill_ with aten::full_like
+}
+
+# In some cases outplace replacement of the operator requires more arguments.
+# We add them manually as we don't want to add unused arguments into the
+# native_functions.yaml
```
Comment (ezyang) [inline tools/autograd/gen_variable_type.py]: Hmm, I don't really understand. Could you explain in more detail? What is "outplace replacement"? What are the "unused arguments"?

Comment (ezyang) [reply tools/autograd/gen_variable_type.py]: I think a before/after description would help.

**Review by ezyang — APPROVED** [review]

File: tools/autograd/gen_variable_type.py (hunk 1 above)
Comment (VitalyFedyunin) [reply tools/autograd/gen_variable_type.py]: I will copy this into the code, so it is searchable by the ops name.

torch.jit.trace have undocumented keyword argument _force_outplace, which force jit to replace functions with outplace variants (for example aten::add_ becomes aten::add).
This replacement implemented in-place with minimum modifications of arguments stack (as it assumes that outplace will have to save arguments as inplace version).
However there are no such substitutions available for aten::fill_ and aten::zero_ operators, as we sever implemented aten::fill and aten::zero, so jit tracing hack replacing aten::zero_ with aten::zeros_like and replacing aten::fill_ with aten::full_like. But as they potentially can have different arguments, we also have to hack into the stack and add missing ones.

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit 51bc068:
* 0/1 flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=51bc068cf1ddaf6ca1cd773c32d2aacbbaa575e5).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

File: tools/autograd/gen_variable_type.py (hunk 1 above)
Comment (ezyang) [reply tools/autograd/gen_variable_type.py]: OK makes sense. I half wonder if it wouldn't have been cleaner to make variants of `full_like` and `zeros_like` which don't take extra factory arguments, so the convention lines up exactly, but this is quibbling.

File: aten/src/ATen/native/TensorFactories.cpp [hunk 2]
```diff
@@ -476,27 +497,41 @@ Tensor& randint_out(
...
 
 Tensor randint_like(
     const Tensor& self,
     int64_t high,
-    const TensorOptions& options) {
+    const TensorOptions& options,
+    c10::optional<c10::MemoryFormat> optional_memory_format) {
+  auto result = at::empty_like(self, options, optional_memory_format);
+  return result.random_(0, high, nullptr);
   return native::randint(high, self.sizes(), nullptr, options);
```
Comment (izdeby) [inline aten/src/ATen/native/TensorFactories.cpp]: @VitalyFedyunin, there are two return statements.

**Pull request closed without merging**